core/trie2/triedb/pathdb/journal.go

```diff
@@ -0,0 +1,303 @@
+	if l == nil {
+		return fmt.Errorf("layer %v not found", root)
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	dbJn := &DBJournal{
+		Version: journalVersion,
+	}
+
+	buf := new(bytes.Buffer)
+	if err := l.journal(buf); err != nil {
+		return err
+	}
+
+	dbJn.EncLayers = buf.Bytes()
+
+	enc, err := encoder.Marshal(dbJn)
+	if err != nil {
+		return err
+	}
+
+	return trieutils.WriteTrieJournal(d.disk, enc)
+}
+
+func (d *Database) loadJournal() (layer, error) {
+	enc, err := trieutils.ReadTrieJournal(d.disk)
+	if err != nil {
+		if !errors.Is(err, db.ErrKeyNotFound) {
+			return nil, err
+		}
+
+		latestID, err := trieutils.ReadPersistedStateID(d.disk)
+		if err != nil {
+			if !errors.Is(err, db.ErrKeyNotFound) {
+				return nil, err
+			}
+		}
+
+		diskRoot := d.getStateRoot()
+		disk := newDiskLayer(&diskRoot, latestID, d, nil, newBuffer(d.config.WriteBufferSize, nil, 0))
+		return disk, nil
+	}
+
+	var journal DBJournal
+	if err := encoder.Unmarshal(enc, &journal); err != nil {
+		return nil, err
+	}
+
+	if journal.Version != journalVersion {
+		return nil, fmt.Errorf("unsupported journal version: %d", journal.Version)
+	}
+
+	head, err := d.loadLayers(journal.EncLayers)
+	if err != nil {
+		return nil, err
+	}
+
+	return head, nil
+}
+
+func (d *Database) loadLayers(enc []byte) (layer, error) {
+	const layerMetadataSize = 9
+
+	var head layer
+	var parent layer
+	for len(enc) > 0 {
+		if len(enc) < layerMetadataSize {
+			return nil, ErrJournalCorrupt
+		}
+		layerType := journalType(enc[0])
+		encLen := binary.BigEndian.Uint64(enc[1:layerMetadataSize])
+
+		if len(enc) < int(encLen) {
+			return nil, ErrJournalCorrupt
+		}
+
+		encLayer := enc[layerMetadataSize : layerMetadataSize+encLen]
+		switch layerType {
+		case diffJournal:
+			var diffJn DiffJournal
+			if err := encoder.Unmarshal(encLayer, &diffJn); err != nil {
+				return nil, err
+			}
+			nodes := new(nodeSet)
+			if err := nodes.decode(diffJn.EncNodeset); err != nil {
+				return nil, err
+			}
+			head = &diffLayer{
+				root:   diffJn.Root,
+				block:  diffJn.Block,
+				nodes:  nodes,
+				parent: parent,
+				id:     parent.stateID() + 1,
+			}
+			parent = head
+		case diskJournal:
+			var diskJn DiskJournal
+			if err := encoder.Unmarshal(encLayer, &diskJn); err != nil {
+				return nil, err
+			}
+
+			latestPersistedID, err := trieutils.ReadPersistedStateID(d.disk)
+			if err != nil {
+				if !errors.Is(err, db.ErrKeyNotFound) {
+					return nil, err
+				}
+			}
+			nodes := new(nodeSet)
+			if err := nodes.decode(diskJn.EncNodeset); err != nil {
+				return nil, err
+			}
+			head = newDiskLayer(
+				&diskJn.Root,
+				diskJn.ID,
+				d,
+				nil,
+				newBuffer(d.config.WriteBufferSize, nodes, diskJn.ID-latestPersistedID),
+			)
+			parent = head
+		default:
+			return nil, fmt.Errorf("unknown journal type: %d", layerType)
+		}
+		enc = enc[layerMetadataSize+encLen:]
+	}
+
+	return head, nil
+}
+
+func (d *Database) getStateRoot() felt.Felt {
```

Comment:
Why is it safe to ignore the errors on this method by simply returning `felt.Zero`?